README
NAME
    Test::API - Test a list of subroutines provided by a module

VERSION
    This documentation describes version 0.002.

SYNOPSIS
         use Test::More tests => 3;
         use Test::API;
     
     require_ok( 'My::Package' );
     
     public_ok ( 'My::Package', @names );
     
     import_ok ( 'My::Package',
             export    => [ 'foo', 'bar' ],
             export_ok => [ 'baz', 'bam' ], 
         );

DESCRIPTION
    This simple test module checks the subroutines provided by a module.
    This is useful for confirming a planned API in testing and ensuring that
    other functions aren't unintentionally included via import.

USAGE
    Note: Subroutines in a package starting with an underscore are ignored.
    Therefore, do not include them in any list of expected subroutines.

  public_ok
       public_ok( $package, @names );

    This function checks that all of the @names provided are available
    within the $package namespace and that only these subroutines are
    available. This means that subroutines imported from other modules will
    cause this test to fail unless they are explicitly included in @names.

  import_ok
       import_ok ( $package, %spec );

    This function checks that $package correctly exports an expected list of
    subroutines and only these subroutines. The %spec generally follows the
    style used by Exporter, but in lower case:

       %spec = (
         export    => [ 'foo', 'bar' ],  # exported automatically
         export_ok => [ 'baz', 'bam' ],  # optional exports
       );

    For "export_ok", the test will check for public functions not listed in
    "export" or "export_ok" that can be imported and will fail if any are
    found.

BUGS
    Please report any bugs or feature requests using the CPAN Request
    Tracker web interface at
    <http://rt.cpan.org/Dist/Display.html?Queue=Test-API>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

SEE ALSO
    *   Test::ClassAPI -- more geared towards class trees with inheritance

AUTHOR
    David A. Golden (DAGOLDEN)

COPYRIGHT AND LICENSE
    Copyright (c) 2009 by David A. Golden. All rights reserved.

    Licensed under Apache License, Version 2.0 (the "License"). You may not
    use this file except in compliance with the License. A copy of the
    License was distributed with this file or you may obtain a copy of the
    License from http://www.apache.org/licenses/LICENSE-2.0

    Files produced as output though the use of this software, shall not be
    considered Derivative Works, but shall be considered the original work
    of the Licensor.

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.



Changes
Revision history for Test-API

{{$NEXT}}

0.010     2018-02-16 22:00:49-05:00 America/New_York

    - No changes from 0.009-TRIAL

0.009     2018-01-22 11:15:13-05:00 America/New_York (TRIAL RELEASE)

    [FIXED]

    - Revised internals for constants/subrefs in the stash (a breaking
      change planned for Perl 5.28)

0.008     2017-04-02 13:30:12-04:00 America/New_York

    - No changes from 0.007-TRIAL

0.007     2017-02-27 09:50:45-05:00 America/New_York (TRIAL RELEASE)

    [TESTS]

    - Fixed tests for perls without '.' in @INC

0.005     2014-04-24 11:09:01-04:00 America/New_York

    [CHANGED]

    - common methods treated as private (e.g. AUTOLOAD, BUILD) are ignored

0.004     2014-03-14 11:43:54+01:00 Europe/Paris

    [ADDED]

    - class_api_ok function tests methods, which may be provided the
      module tested or a superclass (Toby Inkster)

    [CHANGED]

    - dropped dependencies on superclass and Devel::Symdump (Toby Inkster)

0.003     2014-02-27 06:19:07-05:00 America/New_York

    [CHANGED]

    - updated distribution metadata, repo and bugtracker

    - switched to Dist::Zilla for managing this distribution

0.002 Mon Mar  9 18:30:20 EDT 2009

    [ADDED]

    - import_ok() function

0.001 Sun Mar  8 16:38:11 EDT 2009

    - Initial release


t/lib/Child.pm
package t::lib::Child;

use t::lib::Parent;
our @ISA = qw(t::lib::Parent);

sub www { 2 }
sub xxx { 2 }

1;


lib/Test/API.pm
use 5.006;
use strict;
use warnings;

package Test::API;
# ABSTRACT: Test a list of subroutines provided by a module

our $VERSION = '0.011';

use Symbol ();

use Test::Builder::Module 0.86;
our @ISA    = qw/Test::Builder::Module/;
our @EXPORT = qw/public_ok import_ok class_api_ok/;

#--------------------------------------------------------------------------#

sub import_ok ($;@) { ## no critic
    my $package = shift;
    my %spec    = @_;
    for my $key (qw/export export_ok/) {
        $spec{$key} ||= [];
        $spec{$key} = [ $spec{$key} ] unless ref $spec{$key} eq 'ARRAY';
    }
    my $tb = _builder();
    my @errors;
    my %flagged;

    my $label = "importing from $package";

    return 0 unless _check_loaded( $package, $label );

    # test export
    {
        my $test_pkg = *{ Symbol::gensym() }{NAME};
        eval "package $test_pkg; use $package;"; ## no critic
        my ( $ok, $missing, $extra ) = _public_ok( $test_pkg, @{ $spec{export} } );
        if ( !$ok ) {
            push @errors, "not exported: @$missing" if @$missing;
            @flagged{@$missing} = (1) x @$missing if @$missing;
            push @errors, "unexpectedly exported: @$extra" if @$extra;
            @flagged{@$extra} = (1) x @$extra if @$extra;
        }
    }

    # test export_ok
    my @exportable;
    for my $fcn ( _public_fcns($package) ) {
        next if $flagged{$fcn}; # already complaining about this so skip
        next if grep { $fcn eq $_ } @{ $spec{export} }; # exported by default
        my $pkg_name = *{ Symbol::gensym() }{NAME};
        eval "package $pkg_name; use $package '$fcn';"; ## no critic
        my ( $ok, $missing, $extra ) = _public_ok( $pkg_name, $fcn );
        if ($ok) {
            push @exportable, $fcn;
        }
    }
    my ( $missing, $extra ) = _difference( $spec{export_ok}, \@exportable, );
    push @errors, "not optionally exportable: @$missing" if @$missing;
    push @errors, "extra optionally exportable: @$extra" if @$extra;

    # notify of results
    $tb->ok( !@errors, "importing from $package" );
    $tb->diag($_) for @errors;
    return !@errors;
}

#--------------------------------------------------------------------------#

sub public_ok ($;@) { ## no critic
    my ( $package, @expected ) = @_;
    my $tb    = _builder();
    my $label = "public API for $package";

    return 0 unless _check_loaded( $package, $label );

    my ( $ok, $missing, $extra ) = _public_ok( $package, @expected );
    $tb->ok( $ok, $label );
    if ( !$ok ) {
        $tb->diag("missing: @$missing") if @$missing;
        $tb->diag("extra: @$extra")     if @$extra;
    }
    return $ok;
}

#--------------------------------------------------------------------------#

sub class_api_ok ($;@) { ## no critic
    my ( $package, @expected ) = @_;
    my $tb    = _builder();
    my $label = "public API for class $package";

    return 0 unless _check_loaded( $package, $label );

    my ( $ok, $missing, $extra ) = _public_ok( $package, @expected );

    # Call ->can to check if missing methods might be provided
    # by parent classes...
    if ( !$ok ) {
        @$missing = grep { not $package->can($_) } @$missing;
        $ok = not( scalar(@$missing) + scalar(@$extra) );
    }

    $tb->ok( $ok, $label );
    if ( !$ok ) {
        $tb->diag("missing: @$missing") if @$missing;
        $tb->diag("extra: @$extra")     if @$extra;
    }
    return $ok;
}

#--------------------------------------------------------------------------#

sub _builder {
    return __PACKAGE__->builder;
}

#--------------------------------------------------------------------------#

sub _check_loaded {
    my ( $package, $label ) = @_;
    ( my $path = $package ) =~ s{::}{/}g;
    $path .= ".pm";
    if ( $INC{$path} ) {
        return 1;
    }
    else {
        my $tb = _builder();
        local $Test::Builder::Level = $Test::Builder::Level + 1;
        $tb->ok( 0, $label );
        $tb->diag("Module '$package' not loaded");
        return;
    }
}

#--------------------------------------------------------------------------#

sub _difference {
    my ( $array1, $array2 ) = @_;
    my ( %only1, %only2 );
    @only1{@$array1} = (1) x @$array1;
    delete @only1{@$array2};
    @only2{@$array2} = (1) x @$array2;
    delete @only2{@$array1};
    return ( [ sort keys %only1 ], [ sort keys %only2 ] );
}

#--------------------------------------------------------------------------#

# list adapted from Pod::Coverage
my %private = map { ; $_ => 1 } qw(
  import unimport bootstrap

  AUTOLOAD BUILD BUILDARGS CLONE CLONE_SKIP DESTROY DEMOLISH meta

  TIESCALAR TIEARRAY TIEHASH TIEHANDLE

  FETCH STORE UNTIE FETCHSIZE STORESIZE POP PUSH SHIFT UNSHIFT SPLICE
  DELETE EXISTS EXTEND CLEAR FIRSTKEY NEXTKEY PRINT PRINTF WRITE
  READLINE GETC READ CLOSE BINMODE OPEN EOF FILENO SEEK TELL SCALAR

  MODIFY_REF_ATTRIBUTES MODIFY_SCALAR_ATTRIBUTES MODIFY_ARRAY_ATTRIBUTES
  MODIFY_HASH_ATTRIBUTES MODIFY_CODE_ATTRIBUTES MODIFY_GLOB_ATTRIBUTES
  MODIFY_FORMAT_ATTRIBUTES MODIFY_IO_ATTRIBUTES

  FETCH_REF_ATTRIBUTES FETCH_SCALAR_ATTRIBUTES FETCH_ARRAY_ATTRIBUTES
  FETCH_HASH_ATTRIBUTES FETCH_CODE_ATTRIBUTES FETCH_GLOB_ATTRIBUTES
  FETCH_FORMAT_ATTRIBUTES FETCH_IO_ATTRIBUTES
);

sub _public_fcns {
    my ($package) = @_;
    no strict qw(refs);
    my $stash = \%{"$package\::"};
    my @syms;
    for (keys %$stash) {
        push @syms,
             ref \$stash->{$_} eq 'GLOB'
               ? \$stash->{$_}
               : \*{"$package:\:$_"}
    }
    return grep { substr( $_, 0, 1 ) ne '_' && !$private{$_} && $_ !~ /^\(/ }
      map { ( my $f = *$_ ) =~ s/^\*$package\:://; $f }
      grep { defined( *$_{CODE} ) } @syms;
}

#--------------------------------------------------------------------------#

sub _public_ok ($;@) { ## no critic
    my ( $package, @expected ) = @_;
    my @fcns = _public_fcns($package);
    my ( $missing, $extra ) = _difference( \@expected, \@fcns );
    return ( !@$missing && !@$extra, $missing, $extra );
}

1;

__END__

=head1 SYNOPSIS

    use Test::More tests => 3;
    use Test::API;

    require_ok( 'My::Package' );

    public_ok ( 'My::Package', @names );

    import_ok ( 'My::Package',
        export    => [ 'foo', 'bar' ],
        export_ok => [ 'baz', 'bam' ],
    );

    class_api_ok( 'My::Class', @methods );

=head1 DESCRIPTION

This simple test module checks the subroutines provided by a module.  This is
useful for confirming a planned API in testing and ensuring that other
functions aren't unintentionally included via import.

=head1 USAGE

Note: Subroutines starting with an underscore are ignored, as are a number
of other methods not intended to be called directly by end-users.

  import unimport bootstrap

  AUTOLOAD BUILD BUILDARGS CLONE CLONE_SKIP DESTROY DEMOLISH

  TIESCALAR TIEARRAY TIEHASH TIEHANDLE

  FETCH STORE UNTIE FETCHSIZE STORESIZE POP PUSH SHIFT UNSHIFT SPLICE
  DELETE EXISTS EXTEND CLEAR FIRSTKEY NEXTKEY PRINT PRINTF WRITE
  READLINE GETC READ CLOSE BINMODE OPEN EOF FILENO SEEK TELL SCALAR

  MODIFY_REF_ATTRIBUTES MODIFY_SCALAR_ATTRIBUTES MODIFY_ARRAY_ATTRIBUTES
  MODIFY_HASH_ATTRIBUTES MODIFY_CODE_ATTRIBUTES MODIFY_GLOB_ATTRIBUTES
  MODIFY_FORMAT_ATTRIBUTES MODIFY_IO_ATTRIBUTES

  FETCH_REF_ATTRIBUTES FETCH_SCALAR_ATTRIBUTES FETCH_ARRAY_ATTRIBUTES
  FETCH_HASH_ATTRIBUTES FETCH_CODE_ATTRIBUTES FETCH_GLOB_ATTRIBUTES
  FETCH_FORMAT_ATTRIBUTES FETCH_IO_ATTRIBUTES

Therefore, do not include any of these in a list of expected subroutines.

=head2 public_ok

  public_ok( $package, @names );

This function checks that all of the C<@names> provided are available within the
C<$package> namespace and that *only* these subroutines are available.  This
means that subroutines imported from other modules will cause this test to fail
unless they are explicitly included in C<@names>.

=head2 class_api_ok

  class_api_ok( $class, @names );

A variation of C<public_ok> for object-oriented modules. Allows superclasses
to fill in "missing" subroutines, but "extra" methods provided by superclasses
will not cause the test to fail.

=head2 import_ok

  import_ok ( $package, %spec );

This function checks that C<$package> correctly exports an expected list of
subroutines and *only* these subroutines.  The C<%spec> generally follows
the style used by [Exporter], but in lower case:

  %spec = (
    export    => [ 'foo', 'bar' ],  # exported automatically
    export_ok => [ 'baz', 'bam' ],  # optional exports
  );

For C<export_ok>, the test will check for public functions not listed in
C<export> or C<export_ok> that can be imported and will fail if any are found.

=head1 SEE ALSO

=for :list
* L<Test::ClassAPI> -- more geared towards class trees with inheritance

=cut



t/lib/Export.pm
package t::lib::Export;
use strict;
use warnings;

use base 'Exporter';

our @EXPORT = qw/foo/;

sub foo { 1 }

1;



Build.PL
# Copyright (c) 2009 by David Golden. All rights reserved.
# Licensed under Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License was distributed with this file or you may obtain a 
# copy of the License from http://www.apache.org/licenses/LICENSE-2.0

use strict;
use lib 'inc';
eval "require Pod::WikiDoc";
my $class = $@ ? "Module::Build" : "Module::Build::WikiDoc";
eval "require $class";
my $build = $class->new( 
    module_name => 'Test::API',
    dist_author => 'David Golden <dagolden@cpan.org>',
    license => 'apache',
    create_readme => 1,
    create_makefile_pl  => 'traditional',
    requires => {
        'perl' => '5.006',
        'base' => 0,
        'Devel::Symdump' => 2.08,
        'Test::Builder::Module' => 0.86, 
    },
    build_requires => {
        'Carp' => 0,
        'Test::Builder::Tester' => 1.18, # 1.02 in perl 5.008008; 1.18 is new
    },
    meta_add            => { 
        no_index => { 
            directory => [ qw/ examples inc t xt /],
        },
        resources => {
          repository => "http://github.com/dagolden/Test-API/", 
          bugtracker => "http://rt.cpan.org/NoAuth/Bugs.html?Dist=Test-API",
        },
    },
);
$build->create_build_script;



Makefile.PL
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.17;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Test a list of subroutines provided by a module",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "Test-API",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Test::API",
  "PREREQ_PM" => {
    "Symbol" => 0,
    "Test::Builder::Module" => "0.86",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Carp" => 0,
    "Exporter" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::Builder::Tester" => "1.18",
    "Test::More" => 0,
    "constant" => 0,
    "lib" => 0
  },
  "VERSION" => "0.011",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Symbol" => 0,
  "Test::Builder::Module" => "0.86",
  "Test::Builder::Tester" => "1.18",
  "Test::More" => 0,
  "constant" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
